stacker/blueprints/base.py

```diff
@@ -130,6 +131,90 @@ def build_parameter(name, properties):
     return p
 
 
+def validate_variable_type(var_name, var_type, value):
+    """Ensures the value is the correct variable type.
+
+    Args:
+        var_name (str): The name of the defined variable on a blueprint.
+        var_type (type): The type that the value should be.
+        value (obj): The object representing the value provided for the
+            variable
+
+    Returns:
+        object: A python object of type `var_type` based on the provided
+            `value`.
+
+    Raises:
+        ValueError: If the `value` isn't of `var_type` and can't be cast as
+            that type, this is raised.
+    """
+
+    if isinstance(var_type, CFNType):
+        value = CFNParameter(name=var_name, value=value)
+    else:
+        if not isinstance(value, var_type):
+            try:
+                value = var_type(value)
+            except ValueError:
+                raise ValueError("Variable %s must be %s.",
+                                 var_name, var_type)
+    return value
+
+
+def resolve_variable(var_name, var_def, provided_variable, blueprint_name):
+    """Resolve a provided variable value against the variable definition.
+
+    Args:
+        var_name (str): The name of the defined variable on a blueprint.
+        var_def (dict): A dictionary representing the defined variables
+            attributes.
+        provided_variable (:class:`stacker.variables.Variable`): The variable
+            value provided to the blueprint.
+        blueprint_name (str): The name of the blueprint that the variable is
+            being applied to.
+
+    Returns:
+        object: The resolved variable value, could be any python object.
+
+    Raises:
+        MissingVariable: Raised when a variable with no default is not
+            provided a value.
+        UnresolvedVariable: Raised when the provided variable is not already
+            resolved.
+        ValueError: Raised when the value is not the right type and cannot be
+            cast as the correct type. Raised by
+            :func:`stacker.blueprints.base.validate_variable_type`
+    """
+
+    try:
+        var_type = var_def["type"]
+    except KeyError:
+        raise VariableTypeRequired(blueprint_name, var_name)
+
+    if provided_variable:
+        if not provided_variable.resolved:
+            raise UnresolvedVariable(blueprint_name, provided_variable)
+        if provided_variable.value is not None:
+            value = provided_variable.value
+    else:
+        # Variable value not provided, try using the default, if it exists
+        # in the definition
+        try:
+            value = var_def["default"]
+        except KeyError:
+            raise MissingVariable(blueprint_name, var_name)
+
+    # If no validator, return the value as is, otherwise apply validator
+    validator = var_def.get("validator", lambda v: v)
+    value = validator(value)
```

Comment:
I think we should have a preferred exception that validators raise that gets caught here. Then we can include more details about which variable was invalid etc.

Right now if you raise an exception within the validator, you can't tell which variable was being validated.